# Roadmap for the VariantCalling pipeline

## Background

## Get things going

Generate a pipeline that handles raw reads and calculates variants in relation to a given genome assembly.

- Input: 
 - A list of raw reads (should be QCed but the pipe would not know)
 - A reference genome

- Output:
 -  A VCF file with variants


## Communal result storage:

Store results in Kaka

## iRODS

Meta data tag results using iRODS


## Automation

Run QC and this pipe on data reception




